File: android/Detonator/src/main/java/com/iconshot/detonator/helpers/CompareHelper.java
package com.iconshot.detonator.helpers;

import com.iconshot.detonator.element.Style.StyleTransform;

import java.util.List;

public class CompareHelper {
    public static boolean compareObjects(Object a, Object b) {
        if (a == null) {
            return b == null;
        }

        return a.equals(b);
    }

    public static boolean compareColors(Object a, Object b) {
        if (a == null) {
            return b == null;
        }

        if (a instanceof String) {
            return a.equals(b);
        }

        if (a instanceof List) {
            if (!(b instanceof List)) {
                return false;
            }

            List<Double> aList = (List<Double>) a;
            List<Double> bList = (List<Double>) b;

            if (aList.size() != bList.size()) {
                return false;
            }

            int i = 0;

            for (Double aDouble : aList) {
                Double bDouble = bList.get(i);

                if (!CompareHelper.compareObjects(aDouble, bDouble)) {
                    return false;
                }

                i++;
            }

            return true;
        }

        return false;
    }

    public static <T> boolean compareLists(List<T> a, List<T> b) {
        if (a == null) {
            return b == null;
        }

        if (a.size() != b.size()) {
            return false;
        }

        int i = 0;

        for (T aValue : a) {
            T bValue = b.get(i);

            if (!CompareHelper.compareObjects(aValue, bValue)) {
                return false;
            }

            i++;
        }

        return true;
    }

    public static boolean compareTransform(StyleTransform a, StyleTransform b) {
        Object aTranslateX = a != null ? a.translateX : null;
        Object aTranslateY = a != null ? a.translateY : null;
        Object aScale = a != null ? a.scale : null;
        Object aScaleX = a != null ? a.scaleX : null;
        Object aScaleY = a != null ? a.scaleY : null;

        Object bTranslateX = b != null ? b.translateX : null;
        Object bTranslateY = b != null ? b.translateY : null;
        Object bScale = b != null ? b.scale : null;
        Object bScaleX = b != null ? b.scaleX : null;
        Object bScaleY = b != null ? b.scaleY : null;

        return (
                CompareHelper.compareObjects(aTranslateX, bTranslateX) &&
                        CompareHelper.compareObjects(aTranslateY, bTranslateY) &&
                        CompareHelper.compareObjects(aScale, bScale) &&
                        CompareHelper.compareObjects(aScaleX, bScaleX) &&
                        CompareHelper.compareObjects(aScaleY, bScaleY)
                );
    }
}

File: ios/Sources/Detonator/Helper/CompareHelper.swift
class CompareHelper {
    public static func compareStyleSizes(_ a: StyleSize?, _ b: StyleSize?) -> Bool {
        if a == nil {
            return b == nil
        }
        
        switch a {
        case .float(let aFloat):
            switch b {
            case .float(let bFloat):
                return aFloat == bFloat
                
            default:
                return false
            }
            
        case .string(let aString):
            switch b {
            case .string(let bString):
                return aString == bString
                
            default:
                return false
            }
            
        default:
            return false
        }
    }
    
    public static func compareStyleColors(_ a: StyleColor?, _ b: StyleColor?) -> Bool {
        if a == nil {
            return b == nil
        }
        
        switch a {
        case .string(let aString):
            switch b {
            case .string(let bString):
                return aString == bString
            
            default:
                return false
            }
            
        case .array(let aArray):
            switch b {
            case .array(let bArray):
                if aArray.count != bArray.count {
                    return false
                }
                
                for (i, aFloat) in aArray.enumerated() {
                    let bFloat = bArray[i]
                    
                    if aFloat != bFloat {
                        return false
                    }
                }
                
                return true
                
            default:
                return false
            }
            
        default:
            return false
        }
    }
    
    public static func compareArrays<T: Equatable>(_ a: [T]?, _ b: [T]?) -> Bool {
        if a == nil {
            return b == nil
        }
        
        guard let a = a else {
            return false
        }
        
        guard let b = b else {
            return false
        }
        
        if a.count != b.count {
            return false
        }
        
        for (i, aValue) in a.enumerated() {
            let bValue = b[i]
            
            if aValue != bValue {
                return false
            }
        }
        
        return true
    }
    
    public static func compareStyleTransforms(_ a: StyleTransform?, _ b: StyleTransform?) -> Bool {
        let aTranslateX = a != nil ? a!.translateX : nil;
        let aTranslateY = a != nil ? a!.translateY : nil;
        let aScale = a != nil ? a!.scale : nil;
        let aScaleX = a != nil ? a!.scaleX : nil;
        let aScaleY = a != nil ? a!.scaleY : nil;

        let bTranslateX = b != nil ? b!.translateX : nil;
        let bTranslateY = b != nil ? b!.translateY : nil;
        let bScale = b != nil ? b!.scale : nil;
        let bScaleX = b != nil ? b!.scaleX : nil;
        let bScaleY = b != nil ? b!.scaleY : nil;
        
        return (
            CompareHelper.compareStyleSizes(aTranslateX, bTranslateX) &&
            CompareHelper.compareStyleSizes(aTranslateY, bTranslateY) &&
            CompareHelper.compareStyleSizes(aScale, bScale) &&
            CompareHelper.compareStyleSizes(aScaleX, bScaleX) &&
            CompareHelper.compareStyleSizes(aScaleY, bScaleY)
        );
    }
}
